File: Knapsack with Duplicate Items - GFG/knapsack-with-duplicate-items.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG{
    public static void main(String args[])throws IOException
    {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(in.readLine());
        while(t-- > 0){
            String line1[] = in.readLine().trim().split("\\s+");
            int N = Integer.parseInt(line1[0]);
            int W = Integer.parseInt(line1[1]);
            String line2[] = in.readLine().trim().split("\\s+");
            int val[] = new int[N];
            for(int i = 0;i < N;i++)
                val[i] = Integer.parseInt(line2[i]);
            String line3[] = in.readLine().trim().split("\\s+");
            int wt[] = new int[N];
            for(int i = 0;i < N;i++)
                wt[i] = Integer.parseInt(line3[i]);
                
            Solution ob = new Solution();
            System.out.println(ob.knapSack(N, W, val, wt));
        }
    }
}
// } Driver Code Ends


//User function Template for Java

class Solution{
    static int knapSack(int n, int W, int val[], int wt[])
    {
        // code here
        int p[] = new int[W+1];
        for(int j=0; j<=W; j++) p[j]= val[0]*(j/wt[0]); // Infinite supply base scenario ~
        
        for(int i=1; i<n; i++){
            for(int j=0; j<=W; j++){
                int noTake = 0 + p[j], take= Integer.MIN_VALUE;
                
                if(wt[i] <= j) take = val[i] + p[j - wt[i]];
                p[j] = Math.max(take, noTake);
            }
        }
        
        return p[W];
    }
}

File: Knapsack with Duplicate Items - GFG/knapsack-with-duplicate-items.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution{
public:
       int solve(int n,int w,int val[], int wt[],vector<vector<int>>&dp){
           
            if(n==0){
                return (w/wt[n])*val[n];
                
            }
              if(dp[n][w]!=-1)return dp[n][w];
              int ntake=solve(n-1,w,val,wt,dp);
               int take=-1e9;
               if(wt[n]<=w)take=val[n]+solve(n,w-wt[n],val,wt,dp);
                
                 return dp[n][w]= max(ntake,take);
       }
    int knapSack(int N, int W, int val[], int wt[])
    {
          vector<vector<int>>dp(N,vector<int>(W+1,-1));
          return solve(N-1,W,val,wt,dp);
    }
};

//{ Driver Code Starts.

int main(){
    int t;
    cin>>t;
    while(t--){
        int N, W;
        cin>>N>>W;
        int val[N], wt[N];
        for(int i = 0;i < N;i++)
            cin>>val[i];
        for(int i = 0;i < N;i++)
            cin>>wt[i];
        
        Solution ob;
        cout<<ob.knapSack(N, W, val, wt)<<endl;
    }
    return 0;
}
// } Driver Code Ends